api/desecapi/pdns_change_tracker.py

```diff
@@ -0,0 +1,228 @@
+
+    def __init__(self, optimize=True):
+        self.optimize = optimize
+        self._domain_additions = set()
+        self._domain_deletions = set()
+        self._rr_set_additions = {}
+        self._rr_set_modifications = {}
+        self._rr_set_deletions = {}
+
+    def __enter__(self):
+        self._rr_set_modifications = {}
+        self._rr_set_deletions = {}
+        post_save.connect(self._on_rr_post_save, sender=RR)
+        post_delete.connect(self._on_rr_post_delete, sender=RR)
+        post_save.connect(self._on_rr_set_post_save, sender=RRset)
+        post_delete.connect(self._on_rr_set_post_delete, sender=RRset)
+        post_save.connect(self._on_domain_post_save, sender=Domain)
+        post_delete.connect(self._on_domain_post_delete, sender=Domain)
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        if exc_val:
+            return  # TODO proper error treatment
+
+        for domain_name in self._domain_deletions:
+            normalized_name = domain_name.rstrip('.') + '.'
+            pdns_id = normalized_name.translate(str.maketrans({'/': '=2F', '_': '=5F'}))
+
+            assert domain_name not in (self._rr_set_additions.keys()
+                                       | self._rr_set_modifications.keys()
+                                       | self._rr_set_deletions.keys())
+            _pdns_delete(pdns_id)
+
+        for domain_name in self._rr_set_additions.keys() | self._domain_additions:
+            normalized_name = domain_name.rstrip('.') + '.'
+            pdns_id = normalized_name.translate(str.maketrans({'/': '=2F', '_': '=5F'}))
+
+            if domain_name in self._domain_additions:
+                salt = '%016x' % random.randrange(16 ** 16)
+                payload = {'name': normalized_name, 'kind': 'MASTER', 'dnssec': True,
+                           'nsec3param': '1 0 127 %s' % salt, 'nameservers': settings.DEFAULT_NS}
+                _pdns_post('/zones', payload)
+
+            additions = self._rr_set_additions.get(domain_name, set())
+            modifications = self._rr_set_modifications.get(domain_name, set())
+            deletions = self._rr_set_deletions.get(domain_name, set())
+
+            assert additions & deletions == set()
+            assert modifications & deletions == set()
+
+            # Due to disjoint guarantees with `deletions`, we have four types of RR sets:
+            # (1) purely added RR sets
+            # (2) purely modified RR sets
+            # (3) added and modified RR sets
+            # (4) purely deleted RR sets
+
+            # We send RR sets to PDNS if one of the following conditions holds:
+            # (a) RR set was added and has at least one RR
+            # (b) RR set was modified
+            # (c) RR set was deleted
+
+            # Conditions (b) and (c) are already covered in the modifications and deletions list,
+            # we filter the additions list to remove newly-added, but empty RR sets
+            additions -= set([
```

Comment:
You can just use a set comprehension, `additions -= {(...) for ... in additions if not ...}`